### DISCUSSION TITLE
Capire perché un dispositivo mqtt diventa unavailable

### DISCUSSION THREAD

**Discussion by avvalex** [indomus/forum#1785] [GENERALE - Discussioni generiche]
Ciao, secondo voi dove posso guardare per capire come mai uno Shelly UNI ogni tanto diventa unavailable, nonostante continui regolarmente a pubblicare sul broker MQTT (con MQTT Explorer non noto nulla di strano o di anomalo).
Basta riavviare HA e riprende ad andare per qualche giorno.
Qualche idea?

**Comment by snakuzzo** [comment]
Posta la configurazione dell'entità, avrai qualche problema con il LWT (availability topic).

Fai vedere pure come hai configurato la parte MQTT sullo shelly

**Comment by avvalex** [comment]
Non ho ancora creato il package MQTT... Mi interessa capire il discorso del LWT: ossia, se per caso va offline ed entra in servizio l'LWT, se subito dopo lo Shelly riprende ad andare non dovrebbe tornare tutto a funzionare? Perché invece a me quando controllo, pubblica regolarmente ma HA quando lo battezza unavailable smette di considerarlo....
Questa la mia configurazione:

```
- platform: mqtt
  name: "Shelly UNI caldaia T_MAND"
  state_topic: "shellies/shellyuni-caldaia/ext_temperature/0"
  availability_topic: "shellies/shellyuni-caldaia/online"
  value_template: "{{ value }}"
  unit_of_measurement: "°C"
  device_class: temperature
  payload_available: "true"
  payload_not_available: "false"
  
- platform: mqtt
  name: "Shelly UNI caldaia T_RIT"
  state_topic: "shellies/shellyuni-caldaia/ext_temperature/1"
  availability_topic: "shellies/shellyuni-caldaia/online"
  value_template: "{{ value }}"
  unit_of_measurement: "°C"
  device_class: temperature
  payload_available: "true"
  payload_not_available: "false"
  
- platform: mqtt
  name: "Shelly UNI caldaia T_SAN"
  state_topic: "shellies/shellyuni-caldaia/ext_temperature/2"
  availability_topic: "shellies/shellyuni-caldaia/online"
  value_template: "{{ value }}"
  unit_of_measurement: "°C"
  device_class: temperature
  payload_available: "true"
  payload_not_available: "false"
```


<img width="467" alt="Immagine1" src="https://user-images.githubusercontent.com/60694870/180637017-15278801-e94d-4c03-856c-218e1f2e4f32.png">

**Reply by snakuzzo** [reply]
Devi vedere se quando torna online ripubblica l'availability topic con `true` (solitamente viene pubblicato con retain in modo che anche se riavvii HA, HA lo trova sulla coda del broker MQTT).
Se non pubblica `true` lo vedrai sempre `unavailable`

https://indomus.it/formazione/mqtt-cose-e-come-funziona-il-last-will-and-testament-lwt/

**Comment by avvalex** [comment]
In tal caso posso impostare io un'autorizzazione che quando diventa unavailable posti un comando per ricevere l'announce?

**Reply by snakuzzo** [reply]
Potresti farlo, ma non ha senso...deve funzionare tutto senza questa automazione

**Comment by avvalex** [comment]
Infatti. Ho tipo altri 15 Shelly e nessuno ha questo problema.
Però, qualora sia proprio il LWT che rende il dispositivo unavailable, perché magari va offline per un po', e quando lo Shelly torna online non pubblica che è nuovamente attivo... Come dovrei fare?
Cioè, come posso farglielo fare?

**Reply by snakuzzo** [reply]
Io innanzi tutto proverei a resettare il device.
Prova anche a vedere se lato shelly puoi impostare il QoS (non ricordo) mettendolo a 2.
Se dovesse continuare a darti problemi, proverei a segnalerei la cosa a Shelly e magari provare una versione più nuova (una beta magari).
Se proprio non riuscissi a risolvere, potresti fare due cose...

1. Togli l'availability topic dalla configurazione delle entità, così non sarà mai `unavailable`, ma di contro non saprai mai quando il device non è connesso

2. Potresti fare un'automazione utilizzando come trigger lo stato dell'entità e quando diventa `unvailable` fai una publish del topic di announce, forzando la pubblicazione dello stato.

**Reply by snakuzzo** [reply]
P.S. non fare reply in altri thread. Fai sempre rispondi sotto l'ultimo messaggio

**Comment by avvalex** [comment]
Shelly resettato, ma senza risultato.
Il QoS temo non sia il male, non è che perde qualche comando, ma dal momento X diventa e resta unavailable a tempo indefinito, ma sul broker leggo correttamente e regolarmente tutti i messaggi che continuano a essere pubblicati (e sono parecchi, perché quella schedina ha un ADC e 3 sonde di temperatura).
Toglendo l'availability topic non saprei quando non funziona, ma almeno smetterebbe di fare il finto morto?
Io ho provato, dopo un'ora che era unavailable, a pubblicare un comando "announce" e in HA si è risvegliato istantaneamente.
Non ho ancora capito come funziona: un device non risponde, entra in funzione il LWT, ma quando il broker riprende a ricevere qualche messaggio, chi è che deve comunicare che il device funziona nuovamente? Il broker? Il device con un nuovo announce automatico?
Una voltra capito questo:
1) provo a contattare Shelly, se è "colpa" loro;
2) piazzo l'automazione e ho risoto :-)

**Reply by snakuzzo** [reply]
t'avevo detto di fare Reply sotto l'ultimo commento!!!!!!!

**Reply by snakuzzo** [reply]
> Toglendo l'availability topic non saprei quando non funziona, ma almeno smetterebbe di fare il finto morto?

Esatto...sarà sempre "disponibile" su HA e non saprai se non funziona o comunque se non si collega al broker

> ma quando il broker riprende a ricevere qualche messaggio, chi è che deve comunicare che il device funziona nuovamente? Il broker? Il device con un nuovo announce automatico?

Nel momento in cui il lo shelly si collega al broker, lo stesso shelly pubblica il topic di announce con payload `true`...

```
shellies/shellyuni-caldaia/online   true
```

...e comunica al broker il LWT topic (`shellies/shellyuni-caldaia/online   false`). Quando lo shelly si scollega dal broker, dopo alcuni secondi, è lo stesso broker a pubblicare il LWT topic che indica che lo shelly si è disconnesso...

```
shellies/shellyuni-caldaia/online   false
```

Ad ogni modo è "troppo strano" come comportamento. Si tratta di un possibile malfunzionamento software, ma il firmware è sempre lo stesso che utilizzano tutti gli altri shelly al mondo. Non può essere un problema del singolo device...

**Reply by avvalex** [reply]
Ce la posso fare a fare il reply giusto... ;-)
Ipotizziamo che lo Shelly vada in tilt per qualche secondo, e che per questo il broker divulghi il LWT. Lo Shelly riprende la connessione, ma non sa che il broker ha aperto il testamento, per cui credo non si preoccupi di fare un announce per dichiararsi vivo e vegeto.
Quindi, deduco, è il broker che dovrebbe constatare da solo il funzionamento e fare qualcosa di diverso dal riprendere a divulgare le telemetrie? Perché in realtà quando me ne accorgo e ci guardo, lui sta facendo il suo dovere.
È come se HA una volta ricevuto il LWT spegnesse l'ascolto sul sensore, in attesa di un nuovo online true. Lo deve passare qualcuno? Lo Shelly? Il broker? Nessuno?
Tutti gli altri Shelly non hanno mai avuto questo problema, e sono tutti sullo stesso firmware.

**Reply by snakuzzo** [reply]
> Ipotizziamo che lo Shelly vada in tilt per qualche secondo, e che per questo il broker divulghi il LWT. Lo Shelly riprende la connessione, ma non sa che il broker ha aperto il testamento, per cui credo non si preoccupi di fare un announce per dichiararsi vivo e vegeto.

No. Quando lo shelly si collega al broker, indipendentemente da tutto, pubblica il topic di announce con lo stato `true`.
Il broker comunica lo stesso topic lo stato `false` quando non vede più connesso il device.
Se si scollega e ricollega 40 volte, per 40 volte pubblica l'announce con payload `true`

> Quindi, deduco, è il broker che dovrebbe constatare da solo il funzionamento e fare qualcosa di diverso dal riprendere a divulgare le telemetrie? Perché in realtà quando me ne accorgo e ci guardo, lui sta facendo il suo dovere.

stai andando fuori strada. E da quello che dici devi chiarire un attimino meglio il discorso del topic LWT. il broker non è fatto con questa sorta di "IA" per fare queste cose :) ... è un semplice broker. Se io pubblico a manella un topic uguale con payload `pippo` o topic di telemetrie a pene di segugio, quella roba finisce sulla coda del broker e fine. Quindi...non divaghiamo.

> È come se HA una volta ricevuto il LWT spegnesse l'ascolto sul sensore, in attesa di un nuovo online true.

Ti ho già spiegato che non è che HA "spegne l'ascolto". E' che se l'entità ha un availability topic impostato e per qualche ragione il device risulta offline, te lo mostra come `unavailable`. Poi gli altri topic di telemetria possono continuare ad arrivare, ma "GIUSTAMENTE" HA non ti mostra quei valori, visto che l'entità non è disponibile. Ripeto "QUELLO E' SOLO UN BROKER"...non ragionare come se ci fosse chissà quale IA...la logica sta tutta lato HA

> Lo deve passare qualcuno? Lo Shelly? Il broker? Nessuno?

aridaje! :D  ... il topic di availabilty con `true` lo puoi pure pubblicare tu a manella...ma non avrebbe alcun senso....
e se il device fosse offline per davvero ? Non arriverebbe alcuna telemetria...non trovi?
Se proprio vuoi mettere una pezza, scatena la pubblicazione del topic di announce da parte dello shelly tramite un'automazione, anche se continuo a ripeterti che questo comportamento non ha alcun senso.

**Reply by avvalex** [reply]
Quindi il broker non ha notizie dello Shelly, per cui pubblica il testamento.
Lo Shelly per qualche ragione non sa di essere offline, per cui non ripubblica un availability true.
Ma chi è che deve revocare il LWT? Perché in questi casi resterebbe sempre attivo lo stato offline. Ossia, quand'è che uno Shelly dovrebbe decidere di ripubblicare un availability true? Non so se mi spiego....
Io pensavo: il broker non ha notizie, pubblica offline come da ultime volontà; il broker ha nuovamente notizie, ripristina l'online... 
Invece mi pare di capire che sia lo Shelly a doverlo ripetere che è online... Ma se non sapesse di essere finito offline? Strano anche questo con QoS=1....
Cioè, faccio la domanda semplice: se io pubblico a mano un availability offline, come se fosse il broker a pubblicare il LWT per una qualche ragione, chi è che poi ripristinerebbe l'online?

**Reply by snakuzzo** [reply]
> Io pensavo: il broker non ha notizie, pubblica offline come da ultime volontà; il broker ha nuovamente notizie, ripristina l'online...
> Invece mi pare di capire che sia lo Shelly a doverlo ripetere che è online

esatto

> Cioè, faccio la domanda semplice: se io pubblico a mano un availability offline, come se fosse il broker a pubblicare il LWT per una qualche ragione, chi è che poi ripristinerebbe l'online?

nessuno.
Non ricordo se ogni tot minuti/ore lo shelly ripubblica il topic di announce, ma di fatto lo pubblica sicuramente ogni volta che si collega al broker.

P.S. ovviamente parliamo sempre di prodotti Shelly v1. Con quelli v2 (i plus per intenderci) cambia tutta la logica

**Reply by War3333** [reply]
Mi stacco un attimo dalla discussione principale solo per aggiungere che io ho 4 shelly su una trentina che si comportano così.
E' un problema che esiste quando la connessione WiFi degli shelly è molto ballerina quasi al limite.
Ho messo una automazione che ogni 10 minuti chiede a quei 4 shelly l'announce, non c'è molto da fare ho provato ad aumentare il QoS a 2 per fare in modo che quegli shelly avessero certezza di aver raggiunto il broker ma non ha risolto.
Mosquitto pensa che si sia sconnesso e pubblica la morte del device e per qualche motivo anche lo shelly non rimanda mai il topic con il valore true.
